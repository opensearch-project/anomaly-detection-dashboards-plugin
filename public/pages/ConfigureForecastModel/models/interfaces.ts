/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

// Formik values used when creating the model configuration
export interface ModelConfigurationFormikValues {
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  interval: number | undefined;
  windowDelay: number | undefined;
  suggestedSeasonality?: number;
  recencyEmphasis?: number;
  resultIndexMinAge?: number | string;
  resultIndexMinSize?: number | string;
  resultIndexTtl?:number | string;
  flattenCustomResultIndex?: boolean;
  resultIndex?: string;
  horizon?: number;
  history?: number;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  // forecasting has only one feature and we don't need to specify the feature name
  featureName?: string;
  data: number;
}
