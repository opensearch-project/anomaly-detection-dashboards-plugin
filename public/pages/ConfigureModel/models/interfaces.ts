/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  suppressionRules?: RuleFormikValues[];
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  featureName: string;
  data: number;
}

export interface RuleFormikValues {
  featureName: string;
  absoluteThreshold?: number;
  relativeThreshold?: number;
  aboveBelow: string;
}
