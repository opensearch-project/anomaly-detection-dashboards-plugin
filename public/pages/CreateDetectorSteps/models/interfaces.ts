/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { ModelConfigurationFormikValues } from '../../ConfigureModel/models/interfaces';
import { DetectorJobsFormikValues } from '../../DetectorJobs/models/interfaces';

// Formik values used upon creation (includes all fields + those related to historical detector date range)
export interface CreateDetectorFormikValues
  extends DetectorDefinitionFormikValues,
    ModelConfigurationFormikValues,
    DetectorJobsFormikValues {}
