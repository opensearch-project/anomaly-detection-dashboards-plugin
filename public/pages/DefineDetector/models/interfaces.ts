/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { UIFilter } from '../../../models/interfaces';

// Formik values used when creating/editing the detector definition
export interface DetectorDefinitionFormikValues {
  name: string;
  description: string;
  index: { label: string }[];
  resultIndex?: string;
  filters: UIFilter[];
  filterQuery: string;
  timeField: string;
  interval: number;
  windowDelay: number;
  resultIndexMinAge?: number;
  resultIndexMinSize?: number;
  resultIndexTtl?:number;
}
