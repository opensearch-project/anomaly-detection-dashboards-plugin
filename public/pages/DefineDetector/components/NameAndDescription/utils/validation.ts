/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

//TODO:: Check Length, allowed characters.
export const validateDetectorDesc = (
  description: string
): String | undefined => {
  if (description.length > 400) {
    return 'Description Should not exceed 400 characters';
  }
  return undefined;
};
