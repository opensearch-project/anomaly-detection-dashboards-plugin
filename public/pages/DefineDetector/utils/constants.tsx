/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FILTER_TYPES, UIFilter } from '../../../models/interfaces';
import { OPERATORS_MAP } from '../../DefineDetector/components/DataFilterList/utils/constant';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
export const EMPTY_UI_FILTER: UIFilter = {
  filterType: FILTER_TYPES.SIMPLE,
  fieldInfo: [],
  operator: OPERATORS_MAP.IS,
  fieldValue: '',
  query: '',
  label: '',
};

export const testDetectorDefinitionValues: DetectorDefinitionFormikValues = {
  name: 'test',
  description: 'desc',
  index: [],
  filters: [],
  filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
  timeField: 'timestamp',
  interval: 10,
  windowDelay: 1,
  resultIndex: undefined,
};

export const INITIAL_DETECTOR_DEFINITION_VALUES: DetectorDefinitionFormikValues =
  {
    name: '',
    description: '',
    index: [],
    filters: [],
    filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
    timeField: '',
    interval: 10,
    windowDelay: 1,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
    imputationMethod: undefined,
    customImputationValue: undefined
  };
