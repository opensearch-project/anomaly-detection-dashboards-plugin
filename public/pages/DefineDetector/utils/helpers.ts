/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  FILTER_TYPES,
  UNITS,
  Detector,
  UIFilter,
} from '../../../models/interfaces';
import { get, cloneDeep, isEmpty } from 'lodash';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { INITIAL_DETECTOR_DEFINITION_VALUES } from '../../DefineDetector/utils/constants';
import {
  formikToIndices,
  formikToFilterQuery,
} from '../../ReviewAndCreate/utils/helpers';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

export function detectorDefinitionToFormik(
  ad: Detector
): DetectorDefinitionFormikValues {
  const initialValues = cloneDeep(INITIAL_DETECTOR_DEFINITION_VALUES);
  if (isEmpty(ad)) return initialValues;

  return {
    ...initialValues,
    name: ad.name,
    description: ad.description,
    index: [{ label: ad.indices[0] }], // Currently we support only one index
    resultIndex: ad.resultIndex,
    filters: filtersToFormik(ad),
    filterQuery: JSON.stringify(
      get(ad, 'filterQuery', { match_all: {} }),
      null,
      4
    ),
    timeField: ad.timeField,
    interval: get(ad, 'detectionInterval.period.interval', 10),
    windowDelay: get(ad, 'windowDelay.period.interval', 0),
  };
}

export function filtersToFormik(detector: Detector): UIFilter[] {
  // Detectors created or updated using the API will not have metadata - create a custom filter in this case.
  const noMetadata =
    get(detector, 'uiMetadata.filterType') === undefined &&
    get(detector, 'uiMetadata.filters') === undefined;
  const isOldDetector = get(detector, 'uiMetadata.filterType') !== undefined;

  if (noMetadata) {
    return [
      {
        filterType: FILTER_TYPES.CUSTOM,
        query: JSON.stringify(
          get(detector, 'filterQuery', { match_all: {} }),
          null,
          4
        ),
      },
    ];
  }

  const curFilterType = get(detector, 'uiMetadata.filterType');
  const curFilters = get(detector, 'uiMetadata.filters', []);

  // If this is an old detector:
  // If filters is empty: it means it was a custom filter. Convert to a single filter array with a
  // query value equal to the existing filterQuery
  // If a set of simple filters: modify them by injecting the filter type
  // into each existing filter
  if (isOldDetector) {
    if (isEmpty(curFilters)) {
      return [
        {
          filterType: FILTER_TYPES.CUSTOM,
          query: JSON.stringify(
            get(detector, 'filterQuery', { match_all: {} }),
            null,
            4
          ),
        },
      ];
    } else {
      curFilters.forEach((filter: UIFilter) =>
        filter.filterType = curFilterType);
    }
  }
  return curFilters;
}

export function formikToDetectorDefinition(
  values: DetectorDefinitionFormikValues,
  detector: Detector
): Detector {
  let detectorBody = {
    ...detector,
    name: values.name,
    description: values.description,
    indices: formikToIndices(values.index),
    resultIndex: values.resultIndex,
    filterQuery: formikToFilterQuery(values),
    uiMetadata: {
      features: get(detector, 'uiMetadata.features', {}),
      filters: get(values, 'filters', []),
    },
    timeField: values.timeField,
    detectionInterval: {
      period: { interval: values.interval, unit: UNITS.MINUTES },
    },
    windowDelay: {
      period: { interval: values.windowDelay, unit: UNITS.MINUTES },
    },
  } as Detector;

  return detectorBody;
}

export function clearModelConfiguration(ad: Detector): Detector {
  return {
    ...ad,
    featureAttributes: [],
    uiMetadata: {
      ...ad.uiMetadata,
      features: {},
    },
    categoryField: undefined,
    shingleSize: DEFAULT_SHINGLE_SIZE,
  };
}
