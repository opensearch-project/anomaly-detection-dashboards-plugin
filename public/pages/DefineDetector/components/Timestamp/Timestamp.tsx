/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiComboBox, EuiCallOut, EuiSpacer } from '@elastic/eui';
import { Field, FieldProps, FormikProps } from 'formik';
import { debounce, get, isEmpty } from 'lodash';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { AppState } from '../../../../redux/reducers';
import { getPrioritizedIndices } from '../../../../redux/reducers/opensearch';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';
import { sanitizeSearchText } from '../../../utils/helpers';
import { getError, isInvalid, required } from '../../../../utils/utils';
import { DetectorDefinitionFormikValues } from '../../models/interfaces';

interface TimestampProps {
  formikProps: FormikProps<DetectorDefinitionFormikValues>;
}

export function Timestamp(props: TimestampProps) {
  const dispatch = useDispatch();
  const opensearchState = useSelector((state: AppState) => state.opensearch);
  const selectedIndex = get(props, 'formikProps.values.index.0.label', '');
  const isRemoteIndex = selectedIndex.includes(':');
  const [queryText, setQueryText] = useState('');

  const handleSearchChange = debounce(async (searchValue: string) => {
    if (searchValue !== queryText) {
      const sanitizedQuery = sanitizeSearchText(searchValue);
      setQueryText(sanitizedQuery);
      await dispatch(getPrioritizedIndices(sanitizedQuery));
    }
  }, 300);

  const dateFields = Array.from(
    get(opensearchState, 'dataTypes.date', []) as string[]
  );

  const timeStampFieldOptions = isEmpty(dateFields)
    ? []
    : dateFields.map((dateField) => ({ label: dateField }));

  return (
    <ContentPanel
      title="Timestamp"
      titleSize="s"
      subTitle="Select the time field you want to use for the time filter."
    >
      {isRemoteIndex ? (
        <div>
          <EuiCallOut
            title="A remote index is selected, so you need to manually input the time field."
            color="warning"
            iconType="alert"
            size="s"
          />
          <EuiSpacer size="m" />
        </div>
      ) : null}
      <Field name="timeField" validate={required}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            title="Timestamp field"
            hint="Choose the time field you want to use for time filter."
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
          >
            <EuiComboBox
              id="timeField"
              placeholder="Find timestamp"
              options={timeStampFieldOptions}
              onSearchChange={handleSearchChange}
              onCreateOption={(createdOption: string) => {
                const normalizedOptions = createdOption.trim();
                if (!normalizedOptions) return;
                form.setFieldValue('timeField', normalizedOptions);
              }}
              onBlur={() => {
                form.setFieldTouched('timeField', true);
              }}
              onChange={(options) => {
                form.setFieldValue('timeField', get(options, '0.label'));
              }}
              selectedOptions={(field.value && [{ label: field.value }]) || []}
              singleSelection={{ asPlainText: true }}
              isClearable={false}
            />
          </FormattedFormRow>
        )}
      </Field>
    </ContentPanel>
  );
}
