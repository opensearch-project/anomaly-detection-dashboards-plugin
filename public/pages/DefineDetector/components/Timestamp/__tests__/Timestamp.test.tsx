/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render } from '@testing-library/react';
import { Timestamp } from '../Timestamp';
import { DetectorDefinitionFormikValues } from '../../../models/interfaces';
import { CoreServicesContext } from '../../../../../components/CoreServices/CoreServices';
import { FormikProps, Formik } from 'formik';
import { coreServicesMock } from '../../../../../../test/mocks';

import { mockedStore } from '../../../../../redux/utils/testUtils';

const initialState = {
  opensearch: {
    indices: [
      {
        label: 'test-index',
        health: 'green',
      },
    ],
    aliases: [],
    dataTypes: {
      date: ['created_at'],
      date_nanos: ['timestamp'],
    },
    requesting: false,
    searchResult: {},
    errorMessage: '',
  },
};

const values = {
  name: 'test-ad',
  description: 'desc',
  index: [
    {
      label: 'test-index',
      health: 'green',
    },
  ],
  filters: [],
  filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
  timeField: '',
} as DetectorDefinitionFormikValues;

const formikProps = {
  values: { values },
  errors: {},
  touched: {
    index: true,
    name: true,
    timeField: true,
  },
  isSubmitting: false,
  isValidating: false,
  submitCount: 0,
  initialErrors: {},
  initialTouched: {},
  isValid: true,
  dirty: true,
  validateOnBlur: true,
  validateOnChange: true,
  validateOnMount: true,
} as FormikProps<DetectorDefinitionFormikValues>;

describe('<Timestamp /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Provider store={mockedStore(initialState)}>
        <Router>
          <Switch>
            <Route
              render={(props: RouteComponentProps) => (
                <CoreServicesContext.Provider value={coreServicesMock}>
                  <Formik initialValues={values} onSubmit={jest.fn()}>
                    <Timestamp formikProps={formikProps} />
                  </Formik>
                </CoreServicesContext.Provider>
              )}
            />
          </Switch>
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
