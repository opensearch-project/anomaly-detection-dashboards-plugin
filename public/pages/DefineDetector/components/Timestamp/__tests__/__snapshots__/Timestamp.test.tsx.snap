// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timestamp /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding: 20px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small"
        data-test-subj="contentPanelTitle"
      >
        Timestamp
      </h3>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem content-panel-subTitle"
          style="line-height: normal; max-width: 75%;"
        >
          Select the time field you want to use for the time filter.
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        />
      </div>
    </div>
  </div>
  <div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFormRow euiFormRow--compressed"
        hint="Choose the time field you want to use for time filter."
        id="random_html_id-row"
        title="Timestamp field"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Timestamp field
              </p>
              <br />
              <div
                class="euiText euiText--medium sublabel"
                style="max-width: 400px;"
              >
                Choose the time field you want to use for time filter.
              </div>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox euiComboBox--compressed"
            data-test-subj="timestampFilter"
            role="combobox"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--compressed"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap--compressed euiComboBox__inputWrap--noWrap"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <p
                    class="euiComboBoxPlaceholder"
                  >
                    Find timestamp
                  </p>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      id="random_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--small euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
